src/python/modules/ensembl/compara/plugins/TestFiles.py

```diff
@@ -0,0 +1,361 @@
+                if not common_node[dirname]:
+                    del common_node[dirname]
+                if not ref_node[dirname]:
+                    del ref_node[dirname]
+                if not target_node[dirname]:
+                    del target_node[dirname]
+        # Compare the directory trees from their root
+        cmp_trees(self._ref_tree_only, self._target_tree_only, self.common_tree)
+
+    @staticmethod
+    def _get_dir_cmp(path: str) -> Dict:
+        """Returns the directory tree rooted at the given path.
+
+        Note:
+            Files of each folder will be under the key ``.``.
+
+        Args:
+            path: Root path.
+
+        """
+        if os.path.isfile(path):
+            tree_dict = {'.': set(os.path.basename(path))}  # type: Dict
+        else:
+            tree_dict = {'': {}}
+            for dirpath, dirnames, filenames in os.walk(path):
+                dirpath = dirpath.replace(path, '')
+                subtree = reduce(operator.getitem, dirpath.split(os.sep), tree_dict)
+                subtree.update({name: {} for name in dirnames})
+                subtree['.'] = set(filenames)
+            tree_dict = tree_dict['']
+        return tree_dict
+
+    @staticmethod
+    def _prune_tree(root: Dict, paths: Union[str, List] = None, raise_err: bool = False) -> Dict:
+        """Returns a pruned directory tree that only includes the given paths.
+
+        Note:
+            If `paths` is empty, it returns `root`.
+
+        Args:
+            root: Directory tree.
+            paths: Relative directory path(s) to include.
+            raise_err: Raise exception flag.
+
+        Raises:
+            ValueError: if `raise_err` is True and any relative path is not part of the directory tree.
+
+        """
+        if not paths:
+            return root
+        # Make sure paths is a list
+        if isinstance(paths, str):
+            paths = [paths]
+        pruned_tree = {}  # type: Dict
+        for rel_path in paths:
+            # Normalize relative path to ensure reduce() finds it correctly
+            dirnames = os.path.normpath(rel_path).split(os.path.sep)
+            try:
+                tree_node = reduce(operator.getitem, dirnames, root)
+            except KeyError:
+                if raise_err:
+                    # Suppress exception context to display only the ValueError
+                    raise ValueError("Path '{}' not found in the directory tree".format(rel_path)) from None
+                continue
+            # Create/find the path in the pruned tree and attach the subtree
+            pruned_node = pruned_tree
+            for name in dirnames[:-1]:
+                pruned_node = pruned_node.setdefault(name, {})
+            pruned_node[dirnames[-1]] = tree_node
+        return pruned_tree
+
+    @staticmethod
+    def _eval_tree(root: Dict, test_func: Callable) -> List:
+        """Returns the files in the directory tree for which the test function returns False.
```

Comment:
This is counter-intuitive. Usually, such function-based filters select elements when the function returns True